src/vmm/src/builder.rs

```diff
@@ -427,6 +427,27 @@ pub fn build_microvm_from_snapshot(
         vcpu_count,
     )?;
 
+    #[cfg(target_arch = "x86_64")]
+    // Check if we need to scale the TSC.
+    // Checking one VCPU is enough.
+    if vcpus[0]
+        .kvm_vcpu
+        .is_tsc_scaling_required(
+            microvm_state.vcpu_states[0].tsc_khz,
+            &microvm_state.vcpu_states[0].cpuid,
+        )
+        .map_err(|e| MicrovmStateError::IncompatibleState(e.to_string()))
+        .map_err(RestoreMicrovmState)?
+    {
+        // Scale the TSC frequency for all VCPUs.
+        for vcpu in &vcpus {
+            vcpu.kvm_vcpu
+                .set_tsc_khz(microvm_state.vcpu_states[0].tsc_khz)
+                .map_err(|e| MicrovmStateError::IncompatibleState(e.to_string()))
+                .map_err(RestoreMicrovmState)?;
+        }
+    }
+
```

Comment:
This approach leads to some developments that I personally find  both unsafe and hard to follow. For example:

```
    pub fn set_tsc_khz(&self, state_tsc: Option<u32>) -> Result<()> {
        // Extract the TSC freq from the state.
        let snapshot_tsc_freq = match state_tsc {
            Some(freq) => freq,
            None => {
                // The snapshot TSC was already validated.
                unreachable!();
            }
        };
        ...
    }
```
Here relying on the fact that the TSC was already validated is very risky since it's not in the control of the function and not even in the control of the class. `state_tsc` is a parameter that we receive and as far as we know here, it's completely out of our control.

I think this is all generated by the attempt to squeeze the `is_same_model()` check inside the `is_tsc_scaling_required()`, which definitely makes sense initially, but it is not worth the tradeoffs. As an incremental improvement here, just to make the code safer, I would at least separate `is_same_mode()` from `is_tsc_scaling_required()`, changing this code to something like:

```
        if !cpuid::common::is_same_model(&microvm_state.vcpu_states[0].cpuid) {
            let state_tsc = microvm_state.vcpu_states[0]
                .tsc_khz
                .ok_or(MicrovmStateError::IncompatibleState(
                    "TSC Error".to_string(),
                ))
                .map_err(RestoreMicrovmState)?;

            if vcpus[0].kvm_vcpu.is_tsc_scaling_required(state_tsc) {
                for vcpu in &vcpus {
                    vcpu.kvm_vcpu
                        .set_tsc_khz(state_tsc)
                        .map_err(|e| MicrovmStateError::IncompatibleState(e.to_string()))
                        .map_err(RestoreMicrovmState)?;
                }
            }
        }
```

Where `is_tsc_scaling_required()` and `set_tsc_khz()` would also consequently change for the better since they would receive the `state_tsc_freq` as a u32 and not an option.

Cosmetically I guess we can go even further. It would be nice if we could do this operation as part of `KvmVcpu::restore_state()`. If it needs to be done earlier, I think we can find a clean way to do it. The implementation details would vary a lot depending on the requirements. We can discuss options offline.

Reply:
While I don't necessarily see it as unsafe, I understand the concern. There is virtually no cost in refactoring this the way you described (in order to get rid of the `unreachable!` statement), so I tried to do it. Let me know if it's better.

Unfortunately, for the last part of the comment, the checks would ideally happen only once, so this means they can't be done inside `KcmVcpu::restore_state()`. This is what we've been trying to avoid. We considered workarounds for this, such as meddling with the TSC `Option<u32>` inside the state, but they feel clunky. @sandreim what do you think?

Reply:
Failing early and reducing resume cost are good reasons to have only one of the CPU states checked once. This works with the simplifying assumption that the host supports invariant TSC and freq is the same across all CPUs.